### PR TITLE
fix(tianmu): Mathematical Functions of UNSIGNED data returns error.(#1123)

### LINKED ISSUE #1123: bug: when using UNSIGNED data,Mathematical Functions returns error
### Have you read the Contributing Guidelines on issues?

- [X] I have read the [Contributing Guidelines on issues](https://github.com/stoneatom/stonedb/blob/stonedb-5.7-dev/CONTRIBUTING.md).

### Please confirm if bug report does NOT exists already ?

- [X] I confirm there is no existing issue for this

### Describe the problem

when the COLUMN is INT UNSIGNED, the resule set is correct
when the COLUMN is BIGINT UNSIGNED, returns error

```shell
mysql> SELECT 'LN3', LN(CUINTEGER) FROM j1 WHERE j1_key=1;
+-----+--------------------+
| LN3 | LN(CUINTEGER)      |
+-----+--------------------+
| LN3 | 22.180709777219757 |
+-----+--------------------+
1 row in set (0.00 sec)

mysql> SELECT 'LN4', LN(CUBIGINT) FROM j1 WHERE j1_key=1;
+-----+--------------+
| LN4 | LN(CUBIGINT) |
+-----+--------------+
| LN4 |         NULL |
+-----+--------------+
1 row in set, 1 warning (0.00 sec)

mysql> SELECT 'LOG3', LOG(CUINTEGER) FROM j1 WHERE j1_key=1;
+------+--------------------+
| LOG3 | LOG(CUINTEGER)     |
+------+--------------------+
| LOG3 | 22.180709777219757 |
+------+--------------------+
1 row in set (0.00 sec)

mysql> SELECT 'LOG4', LOG(CUBIGINT) FROM j1 WHERE j1_key=1;
+------+---------------+
| LOG4 | LOG(CUBIGINT) |
+------+---------------+
| LOG4 |          NULL |
+------+---------------+
1 row in set, 1 warning (0.00 sec)

mysql> SELECT 'LOG23', LOG2(CUINTEGER) FROM j1 WHERE j1_key=1;
+-------+-------------------+
| LOG23 | LOG2(CUINTEGER)   |
+-------+-------------------+
| LOG23 | 31.99999999899229 |
+-------+-------------------+
1 row in set (0.00 sec)

mysql> SELECT 'LOG24', LOG2(CUBIGINT) FROM j1 WHERE j1_key=1;
+-------+----------------+
| LOG24 | LOG2(CUBIGINT) |
+-------+----------------+
| LOG24 |           NULL |
+-------+----------------+
1 row in set, 1 warning (0.00 sec)
```

### Expected behavior

```shell
mysql> SELECT 'LN3', LN(CUINTEGER) FROM j1 WHERE j1_key=1;
+-----+--------------------+
| LN3 | LN(CUINTEGER)      |
+-----+--------------------+
| LN3 | 22.180709777219757 |
+-----+--------------------+
1 row in set (0.00 sec)

mysql> SELECT 'LN4', LN(CUBIGINT) FROM j1 WHERE j1_key=1;
+-----+------------------+
| LN4 | LN(CUBIGINT)     |
+-----+------------------+
| LN4 | 44.3614195558365 |
+-----+------------------+
1 row in set (0.00 sec)

mysql> SELECT 'LOG3', LOG(CUINTEGER) FROM j1 WHERE j1_key=1;
+------+--------------------+
| LOG3 | LOG(CUINTEGER)     |
+------+--------------------+
| LOG3 | 22.180709777219757 |
+------+--------------------+
1 row in set (0.00 sec)

mysql> SELECT 'LOG4', LOG(CUBIGINT) FROM j1 WHERE j1_key=1;
+------+------------------+
| LOG4 | LOG(CUBIGINT)    |
+------+------------------+
| LOG4 | 44.3614195558365 |
+------+------------------+
1 row in set (0.00 sec)

mysql> SELECT 'LOG23', LOG2(CUINTEGER) FROM j1 WHERE j1_key=1;
+-------+-------------------+
| LOG23 | LOG2(CUINTEGER)   |
+-------+-------------------+
| LOG23 | 31.99999999899229 |
+-------+-------------------+
1 row in set (0.00 sec)

mysql> SELECT 'LOG24', LOG2(CUBIGINT) FROM j1 WHERE j1_key=1;
+-------+----------------+
| LOG24 | LOG2(CUBIGINT) |
+-------+----------------+
| LOG24 |             64 |
+-------+----------------+
1 row in set (0.00 sec)

```

### How To Reproduce

```sql
create table j1 (j1_key int);
insert into j1 values (1),(2),(3),(4),(5),(6),(7),(8),(9),(10),(11),(12),(13),(14),(15),(null);

ALTER TABLE j1 ADD COLUMN CUTINYINT TINYINT UNSIGNED;
ALTER TABLE j1 ADD COLUMN CUSMALLINT SMALLINT UNSIGNED;
ALTER TABLE j1 ADD COLUMN CUINTEGER INT UNSIGNED;
ALTER TABLE j1 ADD COLUMN CUBIGINT BIGINT UNSIGNED;
UPDATE j1 SET CUTINYINT=253;
UPDATE j1 SET CUSMALLINT=65533;
UPDATE j1 SET CUINTEGER=4294967293;
UPDATE j1 SET CUBIGINT=18446744073709551613;

SELECT 'LN3', LN(CUINTEGER) FROM j1 WHERE j1_key=1;
SELECT 'LN4', LN(CUBIGINT) FROM j1 WHERE j1_key=1;

SELECT 'LOG3', LOG(CUINTEGER) FROM j1 WHERE j1_key=1;
SELECT 'LOG4', LOG(CUBIGINT) FROM j1 WHERE j1_key=1;

SELECT 'LOG23', LOG2(CUINTEGER) FROM j1 WHERE j1_key=1;
SELECT 'LOG24', LOG2(CUBIGINT) FROM j1 WHERE j1_key=1;
```

### Environment

```shell
root@ub01:/stonedb57/install# /stonedb57/install/bin/mysqld --version
/stonedb57/install/bin/mysqld  Ver 5.7.36-StoneDB for Linux on x86_64 (build-)
build information as follow:
        Repository address: https://github.com/stoneatom/stonedb.git:stonedb-5.7-dev
        Branch name: stonedb-5.7-dev
        Last commit ID: e395f89e1
        Last commit time: Date:   Tue Dec 6 01:04:24 2022 +0800
        Build time: Date: Tue Dec  6 16:03:58 CST 2022
root@ub01:~# cat /etc/os-release
NAME="Ubuntu"
VERSION="20.04.5 LTS (Focal Fossa)"
ID=ubuntu
ID_LIKE=debian
PRETTY_NAME="Ubuntu 20.04.5 LTS"
VERSION_ID="20.04"
```

### Are you interested in submitting a PR to solve the problem?

- [ ] Yes, I will!

### PR DESCRIPTION
**Pull request by DandreChen**
## Summary about this PR

Issue Number: close #1123 

[summary]
Using double type to accpet unsigned data for mathematics
Add LN(),LOG(),LOG2() mathematical operation test cases to test unsigned type

## Tests Check List

- [ ] Unit test
- [x] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

## Changelog

- [ ] New Feature
- [x] Bug Fix
- [ ] Performance Improvement
- [ ] Build/Testing/CI/CD
- [ ] Documentation
- [ ] Not for changelog (changelog entry is not required)

## Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features

### CODE REVIEW

File: storage/tianmu/core/item_tianmu_field.cpp [hunk 1]
```diff
@@ -66,10 +66,15 @@ void Item_tianmufield::SetType(DataType t) {
   tianmu_type = t;
   switch (tianmu_type.valtype) {
     case DataType::ValueType::VT_FIXED:
-      if (tianmu_type.IsInt())
-        ivalue = new Item_int(static_cast<longlong>(0));
-      else
+      if (tianmu_type.IsInt()) {
+        if (ifield->unsigned_flag) {
```
Comment (RingsC) [inline storage/tianmu/core/item_tianmu_field.cpp]: Pls, refine these `if` statements. Make it more simple and concise. Thxs.

Comment (RingsC) [reply storage/tianmu/core/item_tianmu_field.cpp]: value type is :`VT_FIXED` , but in Item_float, we use `not_fixed_dec`.
```
#define NOT_FIXED_DEC 31

/*
  The longest string my_fcvt can return is 311 + "precision" bytes.
  Here we assume that we never cal my_fcvt() with precision >= NOT_FIXED_DEC
  (+ 1 byte for the terminating '\0').
*/
#define FLOATING_POINT_BUFFER (311 + NOT_FIXED_DEC)

/*
  We want to use the 'e' format in some cases even if we have enough space
  for the 'f' one just to mimic sprintf("%.15g") behavior for large integers,
  and to improve it for numbers < 10^(-4).
  That is, for |x| < 1 we require |x| >= 10^(-15), and for |x| > 1 we require
  it to be integer and be <= 10^DBL_DIG for the 'f' format to be used.
  We don't lose precision, but make cases like "1e200" or "0.00001" look nicer.
*/
#define MAX_DECPT_FOR_F_FORMAT DBL_DIG

/*
  The maximum possible field width for my_gcvt() conversion.
  (DBL_DIG + 2) significant digits + sign + "." + ("e-NNN" or
  MAX_DECPT_FOR_F_FORMAT zeros for cases when |x|<1 and the 'f' format is used).
*/
#define MY_GCVT_MAX_FIELD_WIDTH (DBL_DIG + 4 + MY_MAX(5, MAX_DECPT_FOR_F_FORMAT)) \

```

File: storage/tianmu/core/item_tianmu_field.cpp [hunk 2]
```diff
@@ -118,7 +123,11 @@ void Item_tianmufield::FeedValue() {
   switch (tianmu_type.valtype) {
     case DataType::ValueType::VT_FIXED:
       if (tianmu_type.IsInt())
-        ((Item_int *)ivalue)->value = buf->Get64();
+        if (unsigned_flag) {
+          ((Item_float *)ivalue)->value = static_cast<ulonglong>(buf->Get64());
+        } else {
+          ((Item_int *)ivalue)->value = buf->Get64();
```
Comment (RingsC) [inline storage/tianmu/core/item_tianmu_field.cpp]: `((Item_int *)ivalue)` using c++ style type-cast. Thxs.

**Review by RingsC — CHANGES_REQUESTED** [review]

File: storage/tianmu/core/item_tianmu_field.cpp (hunk 1 above)
Comment (DandreChen) [reply storage/tianmu/core/item_tianmu_field.cpp]: got it. 
Solving problems in a better way.

File: storage/tianmu/core/item_tianmu_field.cpp (hunk 2 above)
Comment (DandreChen) [reply storage/tianmu/core/item_tianmu_field.cpp]: fixed it.

File: mysql-test/suite/tianmu/t/unsigned_type.test [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+UPDATE j2 SET CUTINYINT=253;
+UPDATE j2 SET CUSMALLINT=65533;
+UPDATE j2 SET CUINTEGER=4294967293;
+UPDATE j2 SET CUBIGINT=18446744073709551613;
+
+SHOW CREATE TABLE j2;
+
+# SELECT
+
+SELECT 'LN3_tianmu', LN(CUINTEGER) FROM j1 WHERE j1_key=1;
```
Comment (RingsC) [inline mysql-test/suite/tianmu/t/unsigned_type.test]:
```
Returns the base-2 logarithm of X. If X is less than or equal to 0.0E0, the function returns NULL and a warning “Invalid argument for logarithm” is reported. Returns NULL if X is NULL.
```

**Review by RingsC — CHANGES_REQUESTED** [review]

**Review by RingsC — APPROVED** [review]

**Review by RingsC — APPROVED** [review]

**Pull request closed without merging**